### PR TITLE
[Analyzer] Ip2location, closes #1924

### LINKED ISSUE #1924: [Analyzer] IP2Location.io
## Name
IP2Location.io

## Link
https://www.ip2location.io

## Type of analyzer
Observable -> IP address only

## Why should we use it
It detects IP geolocation and proxy information using API. Free to use.

## Possible implementation
Refer to the documentation section in API page.

### PR DESCRIPTION
**Pull request by aryan-bhokare**
closes #1924

# Description
Added new analyzer that provides a fast and accurate IP Geolocation API tool to determine a user's location and use the geolocation information in different use cases.

## Type of change
- [x] New feature (non-breaking change which adds functionality).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [x] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [] [Advanced-Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [x] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [x] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-share-your-plugin-with-the-community))
    - [] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [x] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-modify-a-plugin).
    - [] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [x] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
- [] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).
- [] If the GUI has been modified:
    - [] I have a provided a screenshot of the result in the PR.
    - [] I have created new frontend tests for the new component or updated existing ones.

### CODE REVIEW

**Comment by aryan-bhokare** [comment]
I followed the following steps.
1. made a branch based on develop.
2. Added the Python Module.
3. Rebuild the project.
4. Added analyzer config.
5. When run without API key the Analyzer runs as expected.
    
image

7. But when we don't use the API key.
    This issue comes up 
image

8. When I run `docker exec -ti intelowl_uwsgi python3 manage.py dumpplugin AnalyzerConfig Ip2location` to dump plugin config. This error comes up:
image

**Comment by mlodic** [comment]
let us check this, could be related to a change we have pushed recently in `develop`

**Comment by 0ssigeno** [comment]
Uff, my bad. Let me make some fixes

**Comment by 0ssigeno** [comment]
Should be fixed.
Please rebase/merge `develop` in this branch

**Comment by aryan-bhokare** [comment]
> Should be fixed. Please rebase/merge `develop` in this branch

migration issue is fixed.

**Comment by aryan-bhokare** [comment]
Only one thing left.
When I am using API key this issue persist.
image

**Comment by aryan-bhokare** [comment]
> Only one thing left. When I am using API key this issue persist. image

am I doing anything wrong?

**Comment by aryan-bhokare** [comment]
> the code seems good, I think you might have still inconstistencies in your databases and maybe you have still applied your old migration instead of this new one

After the migration I once again did the rebuild using commands
`./start test build` and `./start test up -- --build`
but the issue is still there.

File: api_app/analyzers_manager/observable_analyzers/ip2location.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+from api_app.analyzers_manager import classes
+from api_app.analyzers_manager.exceptions import AnalyzerRunException
+from tests.mock_utils import MockUpResponse, if_mock_connections, patch
+
+
+class Ip2location(classes.ObservableAnalyzer):
+    base_url: str = "https://api.ip2location.io/"
+    _api_key_name: str
+    ip2location_api_version: str
```
[the remark this answers is not in the document]
Comment (aryan-bhokare) [reply api_app/analyzers_manager/observable_analyzers/ip2location.py:14]: @mlodic is there a way to remove the previous migration file?

**Comment by 0ssigeno** [comment]
To remove a previous migration, you can use
`docker exec -ti intelowl_uwsgi python3 manage.py migrate APP PREVIOUS_MIGRATION`

In you case should be something like
`docker exec -ti intelowl_uwsgi python3 manage.py migrate analyzers_manager 0059`

**Comment by 0ssigeno** [comment]
Let me see if the error in the CI is framework side, 10 minutes

**Comment by 0ssigeno** [comment]
Yep, I fucked up the serializer.
Instead of having you remake the migration and stuff, I pushed the fix manually on your branch

**Comment by aryan-bhokare** [comment]
> Yep, I fucked up the serializer. Instead of having you remake the migration and stuff, I pushed the fix manually on your branch

I rebased your changes. And rebuilt, the issue (no attribute _api_key_name) still persist maybe it has something to do with my db state. 
Can you suggest something for that?

**Comment by 0ssigeno** [comment]
The issue that you have is that you parameter is called `_api_key_name` instead of `api_key_name`.
The `_` is prepended automatically by the framework because it is a secret.

I'm gonna make the change directly on your branch because it is faster.

**Comment by aryan-bhokare** [comment]
> The issue that you have is that you parameter is called `_api_key_name` instead of `api_key_name`. The `_` is prepended automatically by the framework because it is a secret.
> 
> I'm gonna make the change directly on your branch because it is faster.

Thanks a lot everything works fine now.
I did not know that the framework prepends it with `_`.

**Comment by 0ssigeno** [comment]
That should be written somewhere in the docs if I am not mistaken.
But yeah, with hindsight it is not very user friendly and it does nor provide additional benefits, so I will probably refactor it to be without the `_`

**Review by 0ssigeno — APPROVED** [review]

**Review by mlodic — APPROVED** [review]

**Pull request merged**